spyrmsd/rmsd.py

```diff
@@ -373,3 +378,189 @@ def rmsdwrapper(
+
+    # RMSD is computed by the child process
+    # The results need to be shared with the parent process
+    # https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes
+    result = Value("f")
+    process = Process(
+        target=_rmsd_process,
+        args=(molref, mols, result, symmetry, center, minimize, strip, cache),
+    )
+
+    process.start()
+    process.join(timeout=timeout)
+
+    # Check if the process finished running successfully
+    if process.exitcode != 0:
+        # Actually terminate the process if it still running
+        if process.is_alive():
+            process.terminate()
+
+        return [np.nan] * len(mols)
+    else:
+        # Retrieve the result from the finished job.
+        # Currently MyPy gives an error, it's being worked on: https://github.com/python/typeshed/issues/8799
+        return result.value  # type: ignore[attr-defined]
+
+
+def prmsdwrapper(
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]],
+    mols: Union[molecule.Molecule, List[molecule.Molecule]],
+    num_workers: Union[int, None] = 1,
+    symmetry: bool = True,
+    center: bool = False,
+    minimize: bool = False,
+    strip: bool = True,
+    cache: bool = True,
+    timeout: Optional[float] = None,
+) -> Any:
+    """
+    Compute RMSD between two molecules with a timeout.
+
+    Parameters
+    ----------
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]]
+        Reference molecule
+    mols: Union[molecule.Molecule, List[molecule.Molecule]]
+        Molecules to compare to reference molecule
+    num_workers: int
+        Amount of processor to use for the parallel calculations
+    symmetry: bool, optional
+        Symmetry-corrected RMSD (using graph isomorphism)
+    center: bool, optional
+        Center molecules at origin
+    minimize: bool, optional
+        Minimised RMSD (using the quaternion polynomial method)
+    strip: bool, optional
+        Strip hydrogen atoms
+    timeout: float, optional
+        After how many seconds to stop the RMSD calculations
+
+    Returns
+    -------
+    List[float]
+        RMSDs
+    """
+
+    # Ensure the num_workers is less or equal than the max number of CPUs.
+    # MyPy doesn't like the min() operator since os.cpu_count() can return None in some cases
+    num_workers = min(num_workers, os.cpu_count()) if os.cpu_count() is not None else 1  # type: ignore[type-var]
+
+    # Cast the molecules to lists if they aren't already
+    if not isinstance(molrefs, list):
+        molrefs = [molrefs]
+    if not isinstance(mols, list):
+        mols = [mols]
+
+    # Match the length of the molref
+    if len(molrefs) == 1 and len(molrefs) < len(mols):
+        molrefs = molrefs * len(mols)
+
+    # Ensure molrefs and mols have the same len
+    if not len(molrefs) == len(mols):
+        raise ValueError("The input mol lists have different lengths")
+
+    with ProcessPoolExecutor(max_workers=num_workers) as executor:
```

Comment:
To be honest I am also not too experienced with multiprocessing.. The ideal solution would have been that the timeout variable for the `ProcessPoolExecutor.map()` function would actually terminate the process. However it doesn't, and to my understanding all it does is raise an error, but this only resolves after the process terminates.

Anyway, as for your question regarding the processes, I'm not 100% sure.. I would think that the `executor` spawns a process for each core, and because it runs the _rmsd_timeout, it spawns another process. However from monitoring the processes during test runs, it seems like there are never more processes running that the `num_workers` argument. So maybe the `_rmsd_timeout` runs on the same process, or the original process spawned by the `executor `gets replaced by the `_rmsd_timeout` process? Anyway, I will investigate it in more detail tomorrow (trying to track it using the process IDs from python itself). 
The reason I use this whole approach which spawns individual processes is because this is the only way I found to actually terminate a running process. And then to run it on multiple cores the `ProcessPoolExecutor` seemed like the most straight-forward choice. I will try to look over it again in the weekend, since I completely agree that efficiency and performance should be a big priority.

Reply:
> The ideal solution would have been that the timeout variable for the ProcessPoolExecutor.map() function would actually terminate the process. However it doesn't, and to my understanding all it does is raise an error, but this only resolves after the process terminates.

Yeah, I played around a bit with this and also [`concurrent.futures`](https://docs.python.org/3/library/concurrent.futures.html). With `futures` too it is possible to cancel a task that is waiting to be executed, but not one that is already being executed.

> However from monitoring the processes during test runs, it seems like there are never more processes running that the num_workers argument.

That's interesting. I'll look into it, maybe `strace` will give a bit more insights on what's going on.